ContextManager: wrap context mergeChanges with a fault protection method

kurzee opened the pull request:
Fixes #6748 (I hope)

This is a calculated attempt of a fix for our top crash in Fabric consisting of internal inconsistency issues in CoreDate while trying to save our `writerContext`. Writing this up in detail for posterity.

```
ContextManager.m line 294
-[ContextManager internalSaveContext:]
```
The idea is that looking at the session logs of these crashes, almost every crash occurs after launching the app or returning from the background, such as shortly after `<WordPressAppDelegate:applicationDidBecomeActive:`.

Additionally, almost every stack trace has a Reader stack that follows a call to save the main context which generates subsequent notifications of `NSManagedObjectContextDidSaveNotification` landing on:

- https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift#L1370

And later we land on a different thread for the `writerContext`, which fires off it's own `internalSaveContext` and can crash along the way, generating the crash reports we're seeing:
- https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/WordPress/Classes/Utility/ContextManager.m#L273

But why is this crashing? I was considering that maybe a faulting issue was occurring when the user is returning to the app and Reader object models are faulted, breaking things along the way. @aerych speculated something similar in https://github.com/wordpress-mobile/WordPress-iOS/issues/6757#issuecomment-282364294

Doing a little Googling: http://www.mlsite.net/blog/?p=518, there seems to be old cases in which merging objects after a context save notification can result in issues with object faults. Bingo.

That said, I propose we roll with a new wrapper method for merging objects in contexts, like we do here in Reader, to try and avoid this crash case. I'm not 100% confident this is the fix and there isn't a more sinister issue here, but it's a shot.

To test:
1. Open Reader, open Notifications that end in Reader streams.
2. Background, foreground, kill, restore.
3. Try and find generate a case of merging with faults, I had no luck.

Needs review: @aerych 

Also cc @astralbodies since you dug into this a bit before.

astralbodies commented:
Also lets get @jleandroperez to take a look too.

aerych commented:
Hiya @kurzee 
Code looks spot on and no problems testing.  I wasn't able to reproduce the original issue but its a tricky one so I'm not too surprised.  :shipit: pending @jleandroperez's review.

astralbodies commented:
@aerych I won't have time to review this before cutting 7.1 this morning but I'm okay with us bringing it into the 7.1 branch within a day or so 😄.

Sorry for the delay!

jleandroperez commented on WordPress/Classes/Utility/ContextManager.m:
```diff
@@ -188,6 +188,23 @@ - (BOOL)obtainPermanentIDForObject:(NSManagedObject *)managedObject
+        for (NSManagedObject *object in updates) {
+            if ([object isFault]) {
+                // Force a fault-in of the object's key-values
+                [[context objectWithID:[object objectID]] willAccessValueForKey:nil];
```
Since the **object** itself may belong to another NSManagedObjectContext, it might be a good idea to reload the entity in the **context** that's being merged. I'm not sure what would be the effect of `willAccessValueForKey`, when crossing threads.

Something like this perhaps?

```
       NSArray* updates = [notification.userInfo[NSUpdatedObjectsKey] allObjects];
        for (NSManagedObject* object in updates) {
            NSManagedObject* localMO = [context existingObjectWithID: object.objectID error:nil];
            if (localMO.isFault) {
                [localMO willAccessValueForKey:nil];
            }
        }
```

kurzee commented:
>I've totally missed that snippet, during the Simperium phase out. Added a minor suggestion, regarding reloading the entity.

@jleandroperez good catch! I've added that check.

kurzee commented:
Thanks @aerych, @jleandroperez, @astralbodies! Will cherrypick the commits and bring into 7.1.
